### ISSUE TITLE
Hook that runs after config resources are created/updated

### ISSUE THREAD

**Issue by tomwidmer** [helm/helm#9852]
Use case:

* DB migrations
* Creating users in systems (such as DB, etc.)
* Other one-time (or release-time) jobs that should be run before pods are started/upgraded

Currently, you often want to run a hook job that:

1. Depends on secrets and/or configmaps (that are used by other services in your cluster)
2. Runs before your stateful sets/deployments are started/upgraded

There are some workarounds for this:
1. Create the configmaps/secrets in another pre-install/upgrade hook with a weight set to ensure it is created before the hook job. Then use a standard pre-install/upgrade hook for the actual job.
2. Split your chart into two charts, and deploy the config + upgrade job one, then one with the actual deployments, etc.

The problem with 1 is that the configmaps/secrets aren't managed by Helm, and won't be deleted when you delete the chart. 2. is obviously painful.

I can think of a couple of solutions:
1. Be able to define that you want your hook to run after a particular level of resource has completed creation. e.g. add an annotation: `"helm.sh/hook-run-after": ConfigMap`. This will mean your hook runs after all resources created up to ConfigMap (including Secret, Namespace, etc. - see https://github.com/helm/helm/blob/484d43913f97292648c867b56768775a55e4bba6/pkg/releaseutil/kind_sorter.go) 
2. Just have a new hook type that runs in a fixed position, such as after Service or RoleBindingList. It could be called 'mid-install', 'mid-upgrade' or similar.

**Comment by bacongobbler** [comment]
Most if not all of these operations can be handled using the post-install hook. Have you looked into that?

https://helm.sh/docs/topics/charts_hooks/#writing-a-hook

There are a few examples in community charts that use the post-install hook to run database migrations or one-off jobs.

I will note that most use cases are not that concerned about whether the job runs prior to the pods starting or not. The assumption there being that pods will enter a fail/reboot loop until the migration is complete. So in effect, the post-install hook will run prior to the pods are started. The old pods will continue to serve requests until the migration is complete and the new pods start.

Hope that helps provide some guidance.

**Comment by tomwidmer** [comment]
Yes, post-install is another potential workaround, but there are cases where container startup is very slow, so the fail/reboot loop takes a long time, and there are cases where it won't actually fail anyway, just produce wrong results (and possibly even do this as a 'race' condition, so it will work properly/not work depending on whether it started before the job or not, which you can't even tell that easily)

initContainers is yet another potential workaround (particularly with stateful sets). 

But it seems to me like the majority of pre-install/upgrade hooks actually would run fine after the 'config' type resources have been created already, but that would then allow those hooks to depend on those resources. But that would be a breaking change, hence the suggestion to add a new kind of hook.
